beamer/agent.py

```diff
@@ -31,107 +37,165 @@ def _get_deployment_block(contract_info: dict[str, ContractInfo]) -> BlockNumber
-            fill_manager=target_fill_manager,
+            source_chain_id=source_chain.id,
+            target_chain_id=target_chain.id,
+            request_manager=source_chain.request_manager,
+            fill_manager=target_chain.fill_manager,
             match_checker=self._config.token_match_checker,
             address=self._config.account.address,
             latest_blocks={},
             config=self._config,
-            web3_l1=w3_l1,
+            web3_l1=l1.w3,
             task_pool=self._task_pool,
             claim_request_extension=claim_request_extension,
             l1_resolutions={},
             l1_invalidations={},
         )
+        direction_identifier = f"{direction[0]}->{direction[1]}"
+        event_processor = EventProcessor(context)
+        self._event_monitors[direction[0]].subscribe(event_processor)
+        if source_chain.id != target_chain.id:
+            self._event_monitors[direction[1]].subscribe(event_processor)
+        self._event_processors[direction_identifier] = event_processor
 
-        self._event_processor = EventProcessor(self.context)
-
+    def _init(self) -> None:
+        # Just add one worker, as this effectively serializes the work
+        # This is necessary as we use the account for all resolutions and
+        # would otherwise run into nonce problems
+        self._task_pool = ThreadPoolExecutor(max_workers=1)
+        self._event_processors: dict[str, EventProcessor] = {}
         self._event_monitors: dict[ChainId, EventMonitor] = {}
+        l1 = self._init_l1_chain()
+        chains = self._init_chains()
+        chain_ids = list(chains.keys())
+        if len(chain_ids) == 1:
+            chain_ids.append(chain_ids[0])
+
+        directions = permutations(chain_ids, 2)
 
-        event_monitor_data = {
-            "on_new_events": [self._event_processor.add_events],
-            "on_sync_done": self._event_processor.mark_sync_done,
-            "poll_period": POLL_PERIOD,
-        }
-
-        for chain_id in chain_ids:
-            if chain_id == source_chain_id:
-                chain_related_data = {
-                    "web3": w3_source,
-                    "contracts": (source_request_manager, source_fill_manager),
-                    "deployment_block": _get_deployment_block(source_contracts_info),
-                }
-            else:
-                chain_related_data = {
-                    "web3": w3_target,
-                    "contracts": (target_request_manager, target_fill_manager),
-                    "deployment_block": _get_deployment_block(target_contracts_info),
-                }
-            event_monitor_data.update(chain_related_data)
-            self._event_monitors[chain_id] = EventMonitor(**event_monitor_data)  # type: ignore
+        for direction in set(directions):
+            direction = cast(tuple[ChainId, ChainId], direction)
+            self._setup_direction(direction, chains, l1)
```

Comment:
We are using `set` here to filter duplicates of the form `(x, x)` in case we only have one L2 chain, right?

Would it be better to use `itertools.product` instead of `permutations`?

```python
>>> import itertools as it
>>> list(it.product([1], repeat=2))
[(1, 1)]
>>> list(it.product([1,2], repeat=2))
[(1, 1), (1, 2), (2, 1), (2, 2)]
```
This way, we don't get duplicates and there is no need to filter.

Reply:
just looked this one more time, as i am using chain_id as key in chains, chain_ids should be a list of unique chain ids. if the length is only one then it is adding only in that case a duplicate chain_id to the chain_ids, then permutations should have no duplicate direction in my opinion. before using chain_id i was using name and probably i needed a set here. will just remove set.

Reply:
in normal cases we are not adding same direction like A->A like product is doing in your example it is just exception now if there is really only one chain.

Reply:
> just looked this one more time, as i am using chain_id as key in chains, chain_ids should be a list of unique chain ids. if the length is only one then it is adding only in that case a duplicate chain_id to the chain_ids, then permutations should have no duplicate direction in my opinion.

I do not understand.
Suppose we have this in `chains`: `[1]`. Because the length is `1`, you will duplicate the single chain ID: `[1, 1]`.
And now you give this list to `permutations`, which will generate the following sequence: `[(1, 1), (1, 1)]`.

Could you please explain why that is correct and why it is desired to call `_setup_direction` two times with the same arguments?

Reply:
ok i looked another time :D you are right about permutations, but still product is generating (1,1) and (2,2) as well. Should i keep it ?  We dont have something like this normally in run-your-own-repo